firedrake/meshadapt.py

```diff
@@ -0,0 +1,113 @@
+        """
+        """
+        if isinstance(mesh.topology, fmesh.ExtrudedMeshTopology):
+            raise NotImplementedError("Cannot adapt extruded meshes")
+        if mesh.coordinates.ufl_element().family() != 'Lagrange' \
+           or mesh.coordinates.ufl_element().degree() != 1:
+            raise NotImplementedError("Mesh coordinates must be P1.")
+        if metric.ufl_element().family() != 'Lagrange' \
+           or metric.ufl_element().degree() != 1:
+            raise ValueError("Metric should be a P1 field.")
+        super(AAdaptation, self).__init__(mesh)
+        self.metric = metric
+
+    @utils.cached_property
+    def newmesh(self):
+        """
+        Generates the adapted mesh wrt the metric
+        """
+        plex = self.mesh.topology._plex
+        dim = self.mesh._topological_dimension
+        # PETSc adapt routine expects that the right coordinate section is set
+        # hence the following bloc of code
+        entity_dofs = np.zeros(dim+1, dtype=np.int32)
+        entity_dofs[0] = self.mesh.geometric_dimension()
+        coordSection = plex.createSection([1], entity_dofs, perm=self.mesh.topology._plex_renumbering)
+        dmCoords = plex.getCoordinateDM()
+        dmCoords.setDefaultSection(coordSection)
+
+        with self.mesh.coordinates.dat.vec_ro as coords:
+            plex.setCoordinatesLocal(coords)
+        with self.metric.dat.vec_ro as vec:
+            reordered_metric = dmplex.to_petsc_numbering(vec, self.metric.function_space())
+
+        newplex = plex.adapt(reordered_metric)
+        new_mesh = Mesh(newplex)
+        return new_mesh
+
+    def transfer_solution(self, f, method=None):
+        """
+        Transfers a solution field from the old mesh to the new mesh
+
+        :arg f: function defined on the old mesh that one wants to transfer
+        """
+        # TODO many checks
+        Vnew = functionspace.FunctionSpace(self.newmesh, f.function_space().ufl_element())
+        fnew = function.Function(Vnew)
+
+        if f.ufl_element().family() == 'Lagrange' and f.ufl_element().degree() == 1:
+            fnew.dat.data[:] = f.at(self.newmesh.coordinates.dat.data)
+        elif f.ufl_element().family() == 'Lagrange':
+            degree = f.ufl_element().degree()
+            C = functionspace.VectorFunctionSpace(self.newmesh, 'CG', degree)
+            interp_coordinates = function.Function(C)
+            interp_coordinates.interpolate(self.newmesh.coordinates)
+            fnew.dat.data[:] = f.at(interp_coordinates.dat.data)
+        else:
+            raise NotImplementedError("Can only interpolate CG fields")
+        return fnew
+
+
+def adapt(mesh, metric):
+    """
+    Adapt the mesh to a prescribed metric field.
+
+    :arg mesh: the base mesh to adapt
+    :arg metric: a metric tensor field (a Function of a TensorFunctionSpace)
+
+    :return: a new mesh adapted to the metric
+    """
+    adaptor = AAdaptation(mesh, metric)
+    return adaptor.newmesh
```

Comment:
No. This adapt function only exists for FEniCS compatibility.

We decided that the adaptation operation needs to be a class because it potentially carries internal state (even if the current very simple implementation does not do so). It's not a Poltergeist.